### PR TITLE
feat(query): Procedure support drop if exists and create or replace

### PR DESCRIPTION
**Pull request by TCeason**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

- [x] support drop procedure if exists
- [x] support create or replace procedure

Part of https://github.com/datafuselabs/databend/issues/16501
## Tests

- [x] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/16500)

### CODE REVIEW

**Comment by BohuTANG** [comment]
Also need the syntax `CREATE [ OR REPLACE ] PROCEDURE <name> ...`

File: src/query/management/src/procedure/procedure_mgr.rs [hunk 1]
```diff
@@ -104,6 +111,13 @@ impl ProcedureMgr {
                 vec![ProcedureIdToNameIdent::new_generic(name_ident.tenant(), id).to_string_key()]
             })
             .await?;
+        if dropped.is_none() {
+            if req.if_exists {
+                // Ok
+            } else {
+                return Err(AppError::from(name_ident.unknown_error("drop procedure")).into());
+            }
```
Comment (andylokandy) [inline src/query/management/src/procedure/procedure_mgr.rs:119]:
```suggestion
            if !req.if_exists {
                 return Err(AppError::from(name_ident.unknown_error("drop procedure")).into());
            }
```

**Review by andylokandy — APPROVED** [review]

**Review by drmingdrmer — CHANGES_REQUESTED** [review]
> Reviewed 5 of 7 files at r1, 2 of 2 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/databend/16500)* status: all files reviewed, 3 unresolved discussions (waiting on @sundy-li and @TCeason)

___
*[`src/query/management/src/procedure/procedure_mgr.rs` line 94 at r2](https://reviewable.io/reviews/datafuselabs/databend/16500#-O7Yx6tw0dI5l_12mbGD:-O7Yx6tw0dI5l_12mbGE:b-oloel9) ([raw file](https://github.com/datafuselabs/databend/blob/a6c4a2abd416c3aec278484660953ec8f47f5084/src/query/management/src/procedure/procedure_mgr.rs#L94)):*
> ```XML
>                     } else {
>                         Err(AppError::from(name_ident.unknown_error(func_name!())).into())
>                     }
> ```

IMHO, `CreateOrReplace` does not expect an unknown error. If there is no such record, it should just create one.

The above `create_id_value()` should be already enough. I do not get why this `update_id_value` is requried.



_Code quote:_
```XML
                    let res = self
                        .kv_api
                        .update_id_value(name_ident, meta.clone())
                        .await?;

                    if let Some((id, _meta)) = res {
                        Ok(CreateProcedureReply { procedure_id: *id })
                    } else {
                        Err(AppError::from(name_ident.unknown_error(func_name!())).into())
                    }
```
___
*[`src/query/management/src/procedure/procedure_mgr.rs` line 120 at r2](https://reviewable.io/reviews/datafuselabs/databend/16500#-O7YyC5UEa6PBlt_4C3p:-O7YyC5V85Jdj3ZsMuJc:b4h4cu2) ([raw file](https://github.com/datafuselabs/databend/blob/a6c4a2abd416c3aec278484660953ec8f47f5084/src/query/management/src/procedure/procedure_mgr.rs#L120)):*
> ```XML
>                 return Err(AppError::from(name_ident.unknown_error("drop procedure")).into());
>             }
>         }
> ```

The `if_exists` flag only affects the response to the client; it does not influence operations on the meta-service.

This `if_exists` check should be moved up the call stack to the point where an Error response is actually required, either in the immediate caller or a higher-level caller.

Adn `DropProcedureReq` does not need `if_exists` at all.



_Code quote:_
```XML
        if dropped.is_none() {
            if req.if_exists {
                // Ok
            } else {
                return Err(AppError::from(name_ident.unknown_error("drop procedure")).into());
            }
        }
```

**Comment by TCeason** [comment]
> > Reviewed 5 of 7 files at r1, 2 of 2 files at r2, 1 of 1 files at r3, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/datafuselabs/databend/16500)_ status: all files reviewed, 3 unresolved discussions (waiting on @sundy-li and @TCeason)
> 
> _[`src/query/management/src/procedure/procedure_mgr.rs` line 94 at r2](https://reviewable.io/reviews/datafuselabs/databend/16500#-O7Yx6tw0dI5l_12mbGD:-O7Yx6tw0dI5l_12mbGE:b-oloel9) ([raw file](https://github.com/datafuselabs/databend/blob/a6c4a2abd416c3aec278484660953ec8f47f5084/src/query/management/src/procedure/procedure_mgr.rs#L94)):_
> 
> > ```
> >                     } else {
> >                         Err(AppError::from(name_ident.unknown_error(func_name!())).into())
> >                     }
> > ```
> 
> IMHO, `CreateOrReplace` does not expect an unknown error. If there is no such record, it should just create one.
> 
> The above `create_id_value()` should be already enough. I do not get why this `update_id_value` is requried.
> 
> _Code quote:_
> 
> ```
>                     let res = self
>                         .kv_api
>                         .update_id_value(name_ident, meta.clone())
>                         .await?;
> 
>                     if let Some((id, _meta)) = res {
>                         Ok(CreateProcedureReply { procedure_id: *id })
>                     } else {
>                         Err(AppError::from(name_ident.unknown_error(func_name!())).into())
>                     }
> ```
> 
> _[`src/query/management/src/procedure/procedure_mgr.rs` line 120 at r2](https://reviewable.io/reviews/datafuselabs/databend/16500#-O7YyC5UEa6PBlt_4C3p:-O7YyC5V85Jdj3ZsMuJc:b4h4cu2) ([raw file](https://github.com/datafuselabs/databend/blob/a6c4a2abd416c3aec278484660953ec8f47f5084/src/query/management/src/procedure/procedure_mgr.rs#L120)):_
> 
> > ```
> >                 return Err(AppError::from(name_ident.unknown_error("drop procedure")).into());
> >             }
> >         }
> > ```
> 
> The `if_exists` flag only affects the response to the client; it does not influence operations on the meta-service.
> 
> This `if_exists` check should be moved up the call stack to the point where an Error response is actually required, either in the immediate caller or a higher-level caller.
> 
> Adn `DropProcedureReq` does not need `if_exists` at all.
> 
> _Code quote:_
> 
> ```
>         if dropped.is_none() {
>             if req.if_exists {
>                 // Ok
>             } else {
>                 return Err(AppError::from(name_ident.unknown_error("drop procedure")).into());
>             }
>         }
> ```

Ok, we can refactor these two function. e.g. add update_procedure API. If is replace , call update.

**Review by drmingdrmer — CHANGES_REQUESTED** [review]
> Reviewed 4 of 6 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/databend/16500)* status: 10 of 12 files reviewed, 4 unresolved discussions (waiting on @sundy-li and @TCeason)

___
*[`src/query/service/src/interpreters/interpreter_procedure_create.rs` line 62 at r6](https://reviewable.io/reviews/datafuselabs/databend/16500#-O7dLzygAqHqcwvOlZDY:-O7dLzygAqHqcwvOlZDZ:bfj7wj5) ([raw file](https://github.com/datafuselabs/databend/blob/1fb3b9afaefa73fa86ac96f18a188a2ea8dfc16f/src/query/service/src/interpreters/interpreter_procedure_create.rs#L62)):*
> ```XML
>         let _ = UserApiProvider::instance()
>             .add_procedure(&tenant, create_procedure_req, overriding)
>             .await?;
> ```

The returned result should be consumed: with `CreateOption::CreateIfNotExists`, no error should be returned.



_Code quote:_
```XML
        let overriding = self.plan.create_option.is_overriding();

        let _ = UserApiProvider::instance()
            .add_procedure(&tenant, create_procedure_req, overriding)
            .await?;
```

**Review by drmingdrmer — APPROVED** [review]
The logic looks good to me. And I'd suggest avoiding using ErrorCode if possible.

> Reviewed 11 of 11 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/databend/16500#-:-O7dgc_9ADU4UzUuKcbT:b94y753)* status: all files reviewed, 5 unresolved discussions (waiting on @sundy-li and @TCeason)

___
*[`src/query/service/src/interpreters/interpreter_procedure_create.rs` line 71 at r8](https://reviewable.io/reviews/datafuselabs/databend/16500#-O7df7as34948HB_aaf5:-O7df7as34948HB_aaf6:b-oj2lgo) ([raw file](https://github.com/datafuselabs/databend/blob/d06e9c26b16822b78e8f7f9d5d281aac59847196/src/query/service/src/interpreters/interpreter_procedure_create.rs#L71)):*
> ```XML
>                 return Err(e);
>             }
>         }
> ```

I'd suggest avoid using `ErrorCode` if possible. `ErrorCode` includes all kind of errors that makes it difficult to distinguish(by enumerate every of them) RPC error or logic error like `PROCEDURE_ALREADY_EXISTS`.

If `add_procedure()` returns KVAppError or `Result<Result<..>>`, the error handling like this would be easier.

Such as with `Result<Result<T, LogicError>, MetaError>`, a simple `?` will filter out meta service errors that are not interested.



_Code quote:_
```XML
        if let Err(e) = UserApiProvider::instance()
            .add_procedure(&tenant, create_procedure_req, overriding)
            .await
        {
            if !(self.plan.create_option == CreateOption::CreateIfNotExists
                && e.code() == ErrorCode::PROCEDURE_ALREADY_EXISTS)
            {
                return Err(e);
            }
        }
```

**Comment by TCeason** [comment]
> The logic looks good to me. And I'd suggest avoiding using ErrorCode if possible.
> 
> > Reviewed 11 of 11 files at r8, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/datafuselabs/databend/16500#-:-O7dgc_9ADU4UzUuKcbT:b94y753)_ status: all files reviewed, 5 unresolved discussions (waiting on @sundy-li and @TCeason)
> 
> _[`src/query/service/src/interpreters/interpreter_procedure_create.rs` line 71 at r8](https://reviewable.io/reviews/datafuselabs/databend/16500#-O7df7as34948HB_aaf5:-O7df7as34948HB_aaf6:b-oj2lgo) ([raw file](https://github.com/datafuselabs/databend/blob/d06e9c26b16822b78e8f7f9d5d281aac59847196/src/query/service/src/interpreters/interpreter_procedure_create.rs#L71)):_
> 
> > ```
> >                 return Err(e);
> >             }
> >         }
> > ```
> 
> I'd suggest avoid using `ErrorCode` if possible. `ErrorCode` includes all kind of errors that makes it difficult to distinguish(by enumerate every of them) RPC error or logic error like `PROCEDURE_ALREADY_EXISTS`.
> 
> If `add_procedure()` returns KVAppError or `Result<Result<..>>`, the error handling like this would be easier.
> 
> Such as with `Result<Result<T, LogicError>, MetaError>`, a simple `?` will filter out meta service errors that are not interested.
> 
> _Code quote:_
> 
> ```
>         if let Err(e) = UserApiProvider::instance()
>             .add_procedure(&tenant, create_procedure_req, overriding)
>             .await
>         {
>             if !(self.plan.create_option == CreateOption::CreateIfNotExists
>                 && e.code() == ErrorCode::PROCEDURE_ALREADY_EXISTS)
>             {
>                 return Err(e);
>             }
>         }
> ```

Like this: https://github.com/datafuselabs/databend/pull/16500/commits/72137817193a33bbee4e23b32b441ae2e2486a9d?

Maybe we can refactor all procedure API in next pr that support alter procedure name?

**Comment by drmingdrmer** [comment]
> Like this: [7213781](https://github.com/datafuselabs/databend/commit/72137817193a33bbee4e23b32b441ae2e2486a9d)?
> 
> Maybe we can refactor all procedure API in next pr that support alter procedure name?


Yes, It would be alright with another PR:)

In the code snippet above, the outer `Result` should be a `std::result::Result` with `MetaError` as its error type. The current `ErrorCode` `Result` combines both logical errors and RPC errors, which can lead to confusion for the caller. Separating these error types will provide clearer error handling and improve the API's usability.

**Comment by TCeason** [comment]
> Also need the syntax `CREATE [ OR REPLACE ] PROCEDURE <name> ...`

I think we can merge this pr now. cc @BohuTANG

**Review by drmingdrmer — APPROVED** [review]
> Reviewed 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/databend/16500)* status: all files reviewed, 4 unresolved discussions (waiting on @sundy-li and @TCeason)

**Pull request merged**